In [1]:
from dotenv import load_dotenv

# 加载环境变量
load_dotenv()

True

## 创建文档

数据来源：

CRUD-RAG：大型语言模型检索增强生成的综合中文基准

[IAAR-Shanghai/CRUD-RAG](https://github.com/IAAR-Shanghai/CRUD_RAG/tree/main)



In [2]:
from langchain_community.document_loaders import TextLoader

In [3]:
loaders = [
    TextLoader("data/news1.txt", encoding="utf8"), 
    TextLoader("data/news2.txt", encoding="utf8"),
]
docs = []
for loader in loaders:
    docs.extend(loader.load())

print(len(docs))

2


In [4]:
docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 

In [5]:
print(docs[0])

page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。' metadata={'source': 'data/news1.txt'}


In [6]:
print(len(docs[0].page_content))

811


> ### 京东集团 2023 年第二季度财报要点：
>
> #### 营收与利润
>
> 1. **总营收**：2879 亿元人民币，同比增长 7.6%。
> 2. **净服务收入**：541 亿元人民币，同比增长 30.1%。
> 3. **运营收入**：83 亿元人民币，同比增加 45 亿元人民币。
> 4. **非 GAAP 运营收入**：87 亿元人民币，同比增加 29 亿元人民币。
> 5. **净收入**：归属于普通股股东的净收入为 66 亿元人民币，同比增加 22 亿元人民币。
> 6. **非 GAAP 净收入**：86 亿元人民币，同比增长 31.9%。
>
> #### 利润率与研发费用
>
> 1. **营业利润率**：京东零售未分配项目前的营业利润率为 3.2%，去年同期为 3.4%。
> 2. **研发费用**：41 亿元人民币，同比增长 1.1%。
> 3. **研发费用占比**：占净收入的比例为 1.4%，去年同期为 1.5%。
>
> #### 员工与项目
>
> 1. **员工总数**：截至二季度末，京东体系员工总数超 56 万。
> 2. **京东青年城**：预计投入 60 多亿元建设，提供近 4000 套可拎包入住的员工公寓。
>
> #### 京东物流
>
> 1. **净利润**：经调整后净利润为 8.3 亿元人民币，同比增长 288%。
> 2. **营收**：营收 410 亿元人民币，同比增长 31.2%。

In [7]:
print(docs[1])

page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 2300 亿元。据称，腾讯研发投入和专利数量继续保持高位，行业大模型加速落地驱动产业升级，科技底色更足。在腾讯财报的电话会议中，腾讯高管表示，基础大模型正在顺利推进中，今年下半年会有升级。腾讯表示，AI 是一个增长的催化剂，可以增长效率、还包括对用户服务的质量。同时 AI 还可以更好地促进应用对用户的针对性，还能提升成本效率。截至 6 月，腾讯在全球主要国家和地区专利申请公开总数超过 6.6 万件，专利授权数超过 3.3 万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。据IT之家此前报道，BAT 中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36 氪此前报道称，腾讯自研的“腾讯混元大模型”已经进入应用内测阶段，多位腾讯员工表示，近期收到了内部邮件邀请，可以通过网页或者小程序体验混元大模型。据了解，腾讯内部多个业务已经接入混元大模型进行测试。今年 2 月份，36 氪就称腾讯针对类 ChatGPT 对话式产品已成立“混元助手 (HunyuanAide)”项目组。报道称，混元助手项目组的一号位张正友是腾讯史上最高职级拥有者，早年曾在微软研究院就职，带领视觉团队从事大量学术研究，贡献颇丰。张正友于 2018 年加入腾讯，创建了腾讯机器人实验室 Robotics X，致力推进人机协作的下一代机器人研究。2021 年 1 月 8 日，腾讯宣布张正友成为腾讯首位 17 级研究员 / 杰出科学家。“腾讯混元大模型”的实力究竟如何，还需要等官方正式公开并测试才能得知了。相关阅读：《腾讯首次披露行业大模型进展：不追求参数高，希望成本预算可控》广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。' metadata={'source': 'data/news2.txt'}


In [8]:
print(len(docs[1].page_content))

911


> ### 腾讯2023年Q2财报要点
>
> #### 财务数据
>
> - 二季度研发开支达160亿元。
> - 2018年至今累计研发投入超过2300亿元。
>
> #### 研发与创新
>
> - 研发投入和专利数量持续保持高位。
> - 行业大模型加速落地，驱动产业升级。
> - 基础大模型顺利推进，下半年将有升级。
>
> #### 人工智能
>
> - AI被视为增长催化剂，提升效率和用户服务质量。
> - AI促进应用对用户的针对性，提高成本效率。
>
> #### 专利数据
>
> - 截至6月，全球专利申请公开总数超过6.6万件。
> - 专利授权数超过3.3万件，集中在AI、云技术、大出行、即时通讯等领域。
>
> #### 腾讯混元大模型
>
> - 腾讯混元大模型已进入应用内测阶段。
> - 多个业务已接入混元大模型进行测试。
> - 内部成立“混元助手 (HunyuanAide)”项目组，推进类ChatGPT对话式产品。
> - 项目组由张正友带领，致力于下一代机器人研究。
>
> #### 行业内竞争
>
> - BAT中的其他两家公司已推出通用大模型并开展合作及测试。
> - 马化腾表示腾讯埋头研发，不急于展示半成品。

## 切分文档

In [9]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

In [10]:
text_splitter = RecursiveCharacterTextSplitter(chunk_size=500)
docs = text_splitter.split_documents(docs)

In [11]:
len(docs)

6

In [12]:
docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%

## 导包

In [13]:
from langchain.retrievers.multi_vector import MultiVectorRetriever

In [28]:
from langchain.storage import InMemoryStore
from langchain_community.vectorstores import Chroma
from langchain_community.document_loaders import TextLoader
from langchain_openai import OpenAIEmbeddings
from langchain_text_splitters import RecursiveCharacterTextSplitter

## Summary

### 创建sub_docs —— summary_docs

In [15]:
import uuid

from langchain_core.documents import Document
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

In [16]:
chain = (
    {"doc": lambda x: x.page_content}
    | ChatPromptTemplate.from_template("总结以下文档：\n\n{doc}")
    | ChatOpenAI(max_retries=0)
    | StrOutputParser()
)

In [17]:
docs[0]

Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'})

In [18]:
chain.invoke(docs[0])

'年第二季度，京东集团的营收、净服务收入和运营收入均呈现增长趋势，且各项核心指标均超出市场预期。归属于公司普通股股东的净收入和非 GAAP 净收入也均有较大增长，表现出良好的财务状况。'

In [19]:
summaries = chain.batch(docs, {"max_concurrency": 5})

In [20]:
summaries

['年第二季度，京东集团的营收、净服务收入、运营收入以及归属于公司普通股股东的净收入均呈现增长趋势，表现出良好的财务状况。尤其是净服务收入和非 GAAP 净收入同比增长较为显著，超出市场预期。京东零售在2023年第二季度的营业利润率略有下降，但整体表现仍可观。京东集团在第二季度取得了不错的业绩，展现出强劲的发展势头。',
 '营业利润率下降，但归属于公司普通股股东的净收入和非 GAAP 净收入均有显著增长。研发费用增长略微，但占净收入比例下降。京东体系员工总数超过 56 万，投入建设“京东青年城”并提供员工公寓。京东物流第二季度净利润和营收均有大幅增长。',
 '总结：京东在2023年第二季度取得了良好的业绩，研发费用占比有所下降，员工总数超过56万人。京东物流第二季度净利润增长较快，营收也有不错的增长。未来京东将投入60多亿元建设员工公寓，提供近4000套可拎包入住的公寓。',
 '腾讯在2023年第二季度研发开支达160亿元，累计投入超过2300亿元，专利数量和研发投入保持高位，AI是增长的催化剂，可以提高效率和服务质量，促进应用对用户的针对性，并提升成本效率。腾讯在全球主要国家和地区专利申请超过6.6万件，专利授权数超过3.3万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。其在大模型领域也在顺利推进中，马化腾表示腾讯正在埋头研发，不急于展示半成品。BAT中的其他两家已经拿出自己的通用大模型，并展开合作和测试。',
 '专家，并兼任混元助手项目组组长。腾讯在人工智能领域的研发和专利申请数量显示出其在技术创新和领先地位上的努力和成果。其混元大模型的内测和应用测试也表明了公司在语言处理和对话式产品方面的重要进展。张正友作为腾讯人工智能领域的重要人物，对公司的技术研究和产品开发起着关键作用。腾讯在人工智能、云技术等领域的专利授权数量和业务拓展显示出其在未来技术发展和市场竞争中的重要地位和影响力。',
 'ChatGPT 团队成立了混元助手项目组，项目组的负责人张正友是腾讯史上最高职级拥有者之一。张正友曾在微软研究院工作，带领视觉团队进行学术研究，贡献颇丰。他在腾讯创建了 Robotics X 实验室，致力于推进人机协作的下一代机器人研究。腾讯宣布张正友成为腾讯首位 17 级研究员 / 杰出科学家。腾讯混元大模型的实力还需要官方正式公开并测试才能得知。']

In [21]:
docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%

In [22]:
doc_ids = [str(uuid.uuid4()) for _ in docs]
doc_ids

['a3b83ee5-b602-443a-a052-7dcdedbe6f9e',
 '1eb11e65-b056-4b5a-882b-72ad1c69e14a',
 '4309860b-5f71-4328-86b4-f7e7587a2970',
 'd8726c61-df6f-4144-98a9-fbb8f4758e0d',
 'c7a89159-1ed2-427e-a5c0-08a88858c26d',
 '476543b4-27bb-4e0e-997b-f1fe31571659']

In [23]:
summary_docs = [
    Document(page_content=s, metadata={"doc_id": doc_ids[i]})
    for i, s in enumerate(summaries)
]

In [24]:
summary_docs

[Document(page_content='年第二季度，京东集团的营收、净服务收入、运营收入以及归属于公司普通股股东的净收入均呈现增长趋势，表现出良好的财务状况。尤其是净服务收入和非 GAAP 净收入同比增长较为显著，超出市场预期。京东零售在2023年第二季度的营业利润率略有下降，但整体表现仍可观。京东集团在第二季度取得了不错的业绩，展现出强劲的发展势头。', metadata={'doc_id': 'a3b83ee5-b602-443a-a052-7dcdedbe6f9e'}),
 Document(page_content='营业利润率下降，但归属于公司普通股股东的净收入和非 GAAP 净收入均有显著增长。研发费用增长略微，但占净收入比例下降。京东体系员工总数超过 56 万，投入建设“京东青年城”并提供员工公寓。京东物流第二季度净利润和营收均有大幅增长。', metadata={'doc_id': '1eb11e65-b056-4b5a-882b-72ad1c69e14a'}),
 Document(page_content='总结：京东在2023年第二季度取得了良好的业绩，研发费用占比有所下降，员工总数超过56万人。京东物流第二季度净利润增长较快，营收也有不错的增长。未来京东将投入60多亿元建设员工公寓，提供近4000套可拎包入住的公寓。', metadata={'doc_id': '4309860b-5f71-4328-86b4-f7e7587a2970'}),
 Document(page_content='腾讯在2023年第二季度研发开支达160亿元，累计投入超过2300亿元，专利数量和研发投入保持高位，AI是增长的催化剂，可以提高效率和服务质量，促进应用对用户的针对性，并提升成本效率。腾讯在全球主要国家和地区专利申请超过6.6万件，专利授权数超过3.3万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。其在大模型领域也在顺利推进中，马化腾表示腾讯正在埋头研发，不急于展示半成品。BAT中的其他两家已经拿出自己的通用大模型，并展开合作和测试。', metadata={'doc_id': 'd8726c61-df6f-4144-98a9-fbb8f4758e0d'}),
 Document(page_content='专家，并兼任混元助手项目组

### 创建检索器

In [29]:
# 用于索引子文档的 vectorstore
vectorstore = Chroma(collection_name="summaries", embedding_function=OpenAIEmbeddings())

# 父文档的存储层
store = InMemoryStore()

# 检索器（开始时为空）
retriever = MultiVectorRetriever(
    vectorstore=vectorstore,
    docstore=store,
)

### 存储文档

In [30]:
retriever.vectorstore.add_documents(summary_docs)

['45f73d20-7fb3-4e5d-bb5d-e33fa5368068',
 '83b1e27b-bdf9-4018-bc3e-67ed0f9c5628',
 'ed3c8ea1-8249-4f17-908c-4e5941170e48',
 '9fe3e8d0-e177-4591-ad8c-6025b8c28589',
 '29de6329-4a34-4cd1-ac3a-43e5edab8b5e',
 '9fd8eeab-bd4f-4506-85b0-b529191d6424']

In [31]:
retriever.docstore.mset(list(zip(doc_ids, docs)))

In [32]:
list(zip(doc_ids, docs))

[('a3b83ee5-b602-443a-a052-7dcdedbe6f9e',
  Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'})),
 ('1eb11e65-b056-4b5a-882b-72ad1c69e14a',
  Document(page_content='2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 6

### 检索

#### 测试只使用向量数据库的检索

In [33]:
query = "介绍下腾讯的大模型"

In [34]:
sub_docs = vectorstore.similarity_search(query)

In [35]:
len(sub_docs)

4

In [36]:
sub_docs

[Document(page_content='腾讯在2023年第二季度研发开支达160亿元，累计投入超过2300亿元，专利数量和研发投入保持高位，AI是增长的催化剂，可以提高效率和服务质量，促进应用对用户的针对性，并提升成本效率。腾讯在全球主要国家和地区专利申请超过6.6万件，专利授权数超过3.3万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。其在大模型领域也在顺利推进中，马化腾表示腾讯正在埋头研发，不急于展示半成品。BAT中的其他两家已经拿出自己的通用大模型，并展开合作和测试。', metadata={'doc_id': 'd8726c61-df6f-4144-98a9-fbb8f4758e0d'}),
 Document(page_content='腾讯在2023年第二季度研发开支达160亿元，累计投入超过2300亿元，专利数量和研发投入保持高位，AI是增长的催化剂，可以提高效率和服务质量，促进应用对用户的针对性，并提升成本效率。腾讯在全球主要国家和地区专利申请超过6.6万件，专利授权数超过3.3万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。其在大模型领域也在顺利推进中，马化腾表示腾讯正在埋头研发，不急于展示半成品。BAT中的其他两家已经拿出自己的通用大模型，并展开合作和测试。', metadata={'doc_id': 'd8726c61-df6f-4144-98a9-fbb8f4758e0d'}),
 Document(page_content='专家，并兼任混元助手项目组组长。腾讯在人工智能领域的研发和专利申请数量显示出其在技术创新和领先地位上的努力和成果。其混元大模型的内测和应用测试也表明了公司在语言处理和对话式产品方面的重要进展。张正友作为腾讯人工智能领域的重要人物，对公司的技术研究和产品开发起着关键作用。腾讯在人工智能、云技术等领域的专利授权数量和业务拓展显示出其在未来技术发展和市场竞争中的重要地位和影响力。', metadata={'doc_id': 'c7a89159-1ed2-427e-a5c0-08a88858c26d'}),
 Document(page_content='专家，并兼任混元助手项目组组长。腾讯在人工智能领域的研发和专利申请数量显示出其在技术创新和领先地位上的努力和成果。其混元大模型的内测和应用

#### 测试使用检索器的检索

In [37]:
retrieved_docs = retriever.invoke(query)

In [38]:
len(retrieved_docs)

2

In [39]:
len(retrieved_docs[0].page_content)

440

In [40]:
retrieved_docs

[Document(page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 2300 亿元。据称，腾讯研发投入和专利数量继续保持高位，行业大模型加速落地驱动产业升级，科技底色更足。在腾讯财报的电话会议中，腾讯高管表示，基础大模型正在顺利推进中，今年下半年会有升级。腾讯表示，AI 是一个增长的催化剂，可以增长效率、还包括对用户服务的质量。同时 AI 还可以更好地促进应用对用户的针对性，还能提升成本效率。截至 6 月，腾讯在全球主要国家和地区专利申请公开总数超过 6.6 万件，专利授权数超过 3.3 万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。据IT之家此前报道，BAT 中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36', metadata={'source': 'data/news2.txt'}),
 Document(page_content='6 月，腾讯在全球主要国家和地区专利申请公开总数超过 6.6 万件，专利授权数超过 3.3 万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。据IT之家此前报道，BAT 中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36 氪此前报道称，腾讯自研的“腾讯混元大模型”已经进入应用内测阶段，多位腾讯员工表示，近期收到了内部邮件邀请，可以通过网页或者小程序体验混元大模型。据了解，腾讯内部多个业务已经接入混元大模型进行测试。今年 2 月份，36 氪就称腾讯针对类 ChatGPT 对话式产品已成立“混元助手 (HunyuanAide)”项目组。报道称，混元助手项目组的一号位张正友是腾讯史上最高职级拥有者，早年曾在微软研究院就职，带领视觉团队从事大量学术研究，贡献颇丰。张正友于 2018 年加入腾讯，创建了腾讯机器人实验室 Robotics X，致力推进人机协作的下一代机器人研究。2021 年 1 月 8 日，

In [41]:
query = "京东有多少员工"

retrieved_docs = retriever.invoke(query)

print(len(retrieved_docs))

retrieved_docs

2


[Document(page_content='1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长', metadata={'source': 'data/news1.txt'})]

## Hypothetical Queries

### 创建sub_docs —— question_docs

In [42]:
functions = [
    {
        "name": "hypothetical_questions",
        "description": "生成假设性问题",
        "parameters": {
            "type": "object",
            "properties": {
                "questions": {
                    "type": "array",
                    "items": {"type": "string"},
                },
            },
            "required": ["questions"],
        },
    }
]

In [43]:
from langchain.output_parsers.openai_functions import JsonKeyOutputFunctionsParser

chain = (
    {"doc": lambda x: x.page_content}
    # 只要求生成3个假设性问题，但这个数量是可以调整的
    | ChatPromptTemplate.from_template(
        "生成一个列表，包含正好3个可以用以下文档回答的假设性问题:\n\n{doc}"
    )
    | ChatOpenAI(max_retries=0, model="gpt-4").bind(
        functions=functions, function_call={"name": "hypothetical_questions"}
    )
    | JsonKeyOutputFunctionsParser(key_name="questions")
)


In [44]:
docs[0]

Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'})

In [45]:
chain.invoke(docs[0])

['假设我想知道京东集团2023年第二季度的营收是多少，你能告诉我吗？',
 '假设我想了解京东集团2023年第二季度的净服务收入同比增长了多少，你能告诉我吗？',
 '假设我想知道2023年第二季度京东零售的营业利润率是多少，你能告诉我吗？']

In [46]:
hypothetical_questions = chain.batch(docs, {"max_concurrency": 5})

In [47]:
hypothetical_questions

[['2023年京东集团第二季度的营收是多少？',
  '与2022年相比，京东集团2023年第二季度的净服务收入增长了多少？',
  '京东零售2023年第二季度的营业利润率和2022年第二季度相比有何变化？'],
 ['假设我想知道2023年第二季度的营业利润率和2022年第二季度相比有何变化，那么答案是什么？',
  '如果我想了解2023年第二季度归属于公司普通股股东的净收入和去年同期相比有何变化，我应该如何回答？',
  '假设有人问我2023年第二季度的研发费用和2022年第二季度相比有何变化，我应该如何回答？'],
 ['如果京东的研发费用占净收入比例继续下降，会对公司的发展有何影响？',
  '京东青年城的建设对京东员工的生活有哪些潜在的影响？',
  '京东物流第二季度的经济表现对京东的长期发展有何影响？'],
 ['如果腾讯的研发开支持续增长，那么它将如何影响公司的总开支？',
  '如果腾讯的专利申请没有得到批准，它对公司的研发投入会有什么影响？',
  '如果腾讯的大模型升级失败，它将如何影响公司的AI策略和业务发展？'],
 ['如果腾讯的混元大模型已经成功投入使用，那么它将在哪些领域产生影响？',
  '假设腾讯的专利申请数量在未来几年内持续增长，那么这将如何影响其在全球的市场地位？',
  '如果张正友没有在 2018 年加入腾讯，那么腾讯的机器人研究可能会有什么不同？'],
 ['如果张正友没有加入腾讯，混元助手项目会有什么不同？',
  '如果混元助手项目在 2018 年就开始了，它的进展会如何？',
  '如果腾讯混元大模型的实力超出预期，对行业会有什么影响？']]

In [48]:
# 为每个文档生成唯一标识符
doc_ids = [str(uuid.uuid4()) for _ in docs]

In [49]:
doc_ids

['16714619-cf93-4751-a38b-55ec3e127e87',
 '41e0a3d4-111f-4fa6-ab55-398a9e5e0e0e',
 'b4216264-8aa9-4c6f-b26c-6dcc05e14b3f',
 'f8025be0-69b3-4978-9e2f-ed7645331cec',
 'c02659ba-7bee-4c47-88bd-fd40e8dc428b',
 'c9fd1300-39aa-47b3-8801-f835994d8818']

In [50]:
question_docs = []
for i, question_list in enumerate(hypothetical_questions):
    question_docs.extend(
        [Document(page_content=s, metadata={"doc_id": doc_ids[i]}) for s in question_list]
    )
question_docs

[Document(page_content='2023年京东集团第二季度的营收是多少？', metadata={'doc_id': '16714619-cf93-4751-a38b-55ec3e127e87'}),
 Document(page_content='与2022年相比，京东集团2023年第二季度的净服务收入增长了多少？', metadata={'doc_id': '16714619-cf93-4751-a38b-55ec3e127e87'}),
 Document(page_content='京东零售2023年第二季度的营业利润率和2022年第二季度相比有何变化？', metadata={'doc_id': '16714619-cf93-4751-a38b-55ec3e127e87'}),
 Document(page_content='假设我想知道2023年第二季度的营业利润率和2022年第二季度相比有何变化，那么答案是什么？', metadata={'doc_id': '41e0a3d4-111f-4fa6-ab55-398a9e5e0e0e'}),
 Document(page_content='如果我想了解2023年第二季度归属于公司普通股股东的净收入和去年同期相比有何变化，我应该如何回答？', metadata={'doc_id': '41e0a3d4-111f-4fa6-ab55-398a9e5e0e0e'}),
 Document(page_content='假设有人问我2023年第二季度的研发费用和2022年第二季度相比有何变化，我应该如何回答？', metadata={'doc_id': '41e0a3d4-111f-4fa6-ab55-398a9e5e0e0e'}),
 Document(page_content='如果京东的研发费用占净收入比例继续下降，会对公司的发展有何影响？', metadata={'doc_id': 'b4216264-8aa9-4c6f-b26c-6dcc05e14b3f'}),
 Document(page_content='京东青年城的建设对京东员工的生活有哪些潜在的影响？', metadata={'doc_id': 'b4216264-8aa9-4c6f-b26c-6dcc05e14b3f'}),
 

### 创建检索器

In [51]:
# 用于索引子块的向量存储
vectorstore = Chroma(
    collection_name="hypo-questions", embedding_function=OpenAIEmbeddings()
)

# 父文档的存储层
store = InMemoryStore()


# 检索器（初始为空）
retriever = MultiVectorRetriever(
    vectorstore=vectorstore,
    docstore=store,
)

### 存储文档

In [52]:
retriever.vectorstore.add_documents(question_docs)

['47a281e2-a484-44ff-a293-3957281240b4',
 '399d464f-8834-4a87-92ca-dd14c00803a1',
 '6b9d789b-6c41-4afe-a1ee-0e7e4b210c23',
 '66e1fb97-ba15-4c76-a85c-d3b81a68ea4f',
 '844b6e68-ea35-4bfc-9fca-0cd11508a03e',
 'c8c7703f-1f45-41f2-8194-66c4e920b37d',
 '1df2aa31-60d7-4814-b53c-a6edc4cad118',
 '4f463a90-4c97-473b-9a6f-ce42c5e5586d',
 'fddb2969-91e6-41bb-843e-daadf9640b68',
 '75047486-7faf-42ab-8a8c-8a72ae441868',
 '0bbc14f0-8ac4-4287-aa93-3227a05c4b3d',
 'd81f30ab-34bd-4d22-aa8e-87a92cce13dd',
 '49deac5e-77f3-4d94-89f3-3d1b565e3037',
 '81bbe581-bbe3-41c7-bd6a-00f979ca6e6a',
 'c7f0751d-a13f-46b8-9213-ba462436b951',
 '44b6ed36-1897-48b5-b45c-87347ab25902',
 '47e3d45a-490f-415e-9e2e-a7edf69a9871',
 'e0ab04ec-0890-42ae-85f4-0f7c76abbfb7']

In [53]:
retriever.docstore.mset(list(zip(doc_ids, docs)))

In [54]:
list(zip(doc_ids, docs))

[('16714619-cf93-4751-a38b-55ec3e127e87',
  Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'})),
 ('41e0a3d4-111f-4fa6-ab55-398a9e5e0e0e',
  Document(page_content='2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 6

### 检索

#### 测试只使用向量数据库的检索

In [55]:
query

'京东有多少员工'

In [56]:
sub_docs = vectorstore.similarity_search(query)

In [57]:
sub_docs

[Document(page_content='京东青年城的建设对京东员工的生活有哪些潜在的影响？', metadata={'doc_id': 'b4216264-8aa9-4c6f-b26c-6dcc05e14b3f'}),
 Document(page_content='2023年京东集团第二季度的营收是多少？', metadata={'doc_id': '16714619-cf93-4751-a38b-55ec3e127e87'}),
 Document(page_content='京东零售2023年第二季度的营业利润率和2022年第二季度相比有何变化？', metadata={'doc_id': '16714619-cf93-4751-a38b-55ec3e127e87'}),
 Document(page_content='京东物流第二季度的经济表现对京东的长期发展有何影响？', metadata={'doc_id': 'b4216264-8aa9-4c6f-b26c-6dcc05e14b3f'})]

#### 测试使用检索器的检索

In [58]:
retrieved_docs = retriever.invoke(query)

In [59]:
retrieved_docs

[Document(page_content='1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt'})]